Vue googlevision game (#62)
auth0-blog/guest-writer: articles/2019-11-18-title.markdown

  domain,
  clientId,
  onRedirectCallback: appState => {
    router.push(
      appState && appState.targetUrl
        ? appState.targetUrl
        : window.location.pathname
    );
  }
});

Vue.config.productionTip = false;

new Vue({
  router,
  render: h => h(App)
}).$mount("#app");
```

This file loads in the Auth0 configuration, the `router` and authentication service plugin. The plugin is set up for use within the application using `Vue.use` with the required parameters, the `router` is also loaded into the root instance of the application.

**Note:** If you run this project at this point (or at any other point in the course of this article) and you run into errors relating to `core-js`, simply run `npm install core-js` to fix the issue.

## Building the Emoji game

Now to the main action, this is where you tie everything together. You will be writing code to do the following:

1. Load and display in a select list the camera input options available on the system
2. Display the feed of the currently selected camera input
3. A **Play** button to initiate a new game
4. A "Skip Emoji" button to move to a different emoji is the player can't predict the current one

> 5. Display an object Emoji (no smileys)

Author: This is because we don't want to include the prediction of facial expressions, only objects.

},
logout() {
  this.$auth.logout({
    returnTo: window.location.origin
  });
},

```

### The media stream functions

The `gotDevices`, `getStream` and `gotStream` functions are responsible for handling the collation of video sources and display of the selected video source feed in an HTML video tag for the user to capture.

```javascript
gotDevices(deviceInfos) {
  for (var i = 0; i !== deviceInfos.length; ++i) {
    var deviceInfo = deviceInfos[i];

    if (deviceInfo.kind === "videoinput") {
      let option = {};

      option.id = deviceInfo.deviceId;
      option.text = deviceInfo.label || "camera " + (i - 1);

      this.videoSources.push(option);
    }
  }
},
getStream() {
  if (window.stream) {
    window.stream.getTracks().forEach(function(track) {
      track.stop();
    });
  }

  var constraints = {
    video: {
      deviceId: { exact: this.selectedSource }
    }
  };

  navigator.mediaDevices
    .getUserMedia(constraints)
    .then(this.gotStream)
    .catch(this.handleError);
},
gotStream(stream) {
  this.$refs.video1.srcObject = stream;
}

```

### The Game play functions

These are the functions responsible for handing the game playing activities.

> 1. `playGame`: Resets the game for a user to start a new game.

Reviewer: Again, check for punctuation for each numbered item. Some have periods, some do not.

2. `skipEmoji`: Skips the currently displayed emoji and shows a new one to predict

> 3. `captureImage`: Draws the current frame in the video element unto a canvas element and converts the canvas image to a data URL that is then loaded to the screen using an image tag.

Reviewer: ```suggestion
3. `captureImage`: Draws the current frame in the video element into a canvas element and converts the canvas image to a data URL that is then loaded to the screen using an image tag.
```

4. `predictImage`: Makes a request to the Google Vision API with the captured image to return prediction results, then tries to find a match between the predicted values and the emoji label. If a match exists, the user scores points, if not, the user loses points.

> 5. `switchEmoji`: Changes the currently displayed emoji

Reviewer: ```suggestion
5. `switchEmoji`: Changes the emoji currently displayed.
```

6. `setTimer`: Starts the timer countdown for a game round.

> 7. `resetTimer`: Resets the timer back to the starting point

Reviewer: ```suggestion
7. `resetTimer`: Resets the timer back to zero.
```

8. `endGame`: Ends the game
9. `startNewGame`: Start/Restarts the game

> 10. `showModal`: Display the notifications modal

Reviewer: ```suggestion
10. `showModal`: Displays the notifications modal
```

### The game page template

In the game page template, a top navigation bar displays the login/logout buttons based on the authentication state.

Next, there are two sections, one for the non-authenticated state and the other for the authenticated state. These two sections are displayed conditionally based on the authenticated state.

The non-authenticated section consists of the game homepage listing the game instructions and login button.

> The authenticated section consists of the actual game. A select list of video input sources, a button to start a new game and another button to skip an emoji while in play. Also, there is a video display of the feed of the currently selected video source, a capture button, the button that triggers the prediction, the emoji, score and timer displays.

Reviewer: ```suggestion
The authenticated section consists of the actual game. A select list of video input sources, a button to start a new game and another button to skip an emoji while in play. Also, there is a video display of the feed of the video source selected, a capture button, the button that triggers the prediction, the emoji, score, and timer displays.
```

## Running the Application

> Now its time to put all the work that has been done so far to test. Run the app by running the following command:

Reviewer: ```suggestion
Now it's time to put all the work that has been done so far to test. Run the app by running the following command:
```

`npm run serve`

If all goes smoothly, you should see the login screen as shown below:

![Login Page](images/login-page.png "Login Page - Emoji Game")

Now click the login button. You will see the Auth0 login prompt where you can log in with your email or Gmail/Facebook account.

Once the login process is complete, you will be redirected back to the app where the game page is then displayed. Select a camera source and hit the **Play** button. You should see a page similar to the one below:

![Game Page](images/game-page.png "Game Page - Emoji Game")

> And there you have it, a simple emoji prediction game using Auth0, Vue and the Google Cloud Vision API.

Reviewer: ```suggestion
And there you have it, a simple emoji prediction game using Auth0, Vue, and the Google Cloud Vision API.
```

          emoji.category.includes("Objects") &&
          emoji.char.charCodeAt(0) != 55358
        );
      });
    }
  },
  methods: {
    // Log the user in
    login() {
      this.$auth.loginWithRedirect();
    },
    // Log the user out
    logout() {
      this.$auth.logout({
        returnTo: window.location.origin
      });
    },
    gotDevices(deviceInfos) {
      for (var i = 0; i !== deviceInfos.length; ++i) {
        var deviceInfo = deviceInfos[i];

        if (deviceInfo.kind === "videoinput") {
          let option = {};

          option.id = deviceInfo.deviceId;
          option.text = deviceInfo.label || "camera " + (i - 1);

          this.videoSources.push(option);
        }
      }

    },
    getStream() {
      if (window.stream) {
        window.stream.getTracks().forEach(function(track) {
          track.stop();
        });
      }

      var constraints = {
        video: {
          deviceId: { exact: this.selectedSource }
        }
      };

      navigator.mediaDevices
        .getUserMedia(constraints)
        .then(this.gotStream)
        .catch(this.handleError);
    },
    gotStream(stream) {

      this.$refs.video1.srcObject = stream;
    },

>     handleError(error) {

Reviewer: This needs to be 

 handleError(error) {
      this.showModal(
          "Error",
          error.message
      );
    },

      });

      try {

>         // If the user is returning to the app after authentication..

Reviewer: Let's remove one of the periods

      imageURL: null,
      predictingImage: false,
      gCloudVisionUrl:

>         "https://vision.googleapis.com/v1/images:annotate?key=AIzaSyCk-S9i5xMbmiEAIVPtSC8VlyecH1rY8Uo",

Reviewer: Let's go back to "YOUR_GOOGLE_PROJECT_API_KEY"
